File: .__old/abc405/d.cc
#include <iostream>
#include <vector>
#include <queue>
#include <map>
using namespace std;

int H, W;
vector<vector<char>> Grid, ans;

vector<pair<int, int>> Directions = {
    {1, 0}, {0, 1}, {-1, 0}, {0, -1}
};

map<pair<int, int>, char> DirectionStr = {
    {{1, 0}, '^'},
    {{0, 1}, '<'},
    {{-1, 0}, 'v'},
    {{0, -1}, '>'}
};

char strByDir(pair<int, int> a, pair<int, int> b) {
    int dh = a.first - b.first;
    int dw = a.second - b.second;
    return DirectionStr[{dh, dw}];
}

vector<pair<int, int>> get_next(pair<int, int> pos, pair<int, int> from) {
    vector<pair<int, int>> retval;
    int h = pos.first, w = pos.second;

    for (auto [dh, dw] : Directions) {
        int nh = h + dh, nw = w + dw;
        if (nh < 0 || nh >= H || nw < 0 || nw >= W) continue;
        if (nh == from.first && nw == from.second) continue;
        if (Grid[nh][nw] != '#' && Grid[nh][nw] != 'E') {
            retval.emplace_back(nh, nw);
        }
    }
    return retval;
}

int main() {
    cin >> H >> W;
    Grid.resize(H, vector<char>(W));
    ans.resize(H, vector<char>(W));

    vector<pair<int, int>> start;

    for (int h = 0; h < H; ++h) {
        for (int w = 0; w < W; ++w) {
            cin >> Grid[h][w];
            ans[h][w] = Grid[h][w];
            if (Grid[h][w] == 'E') {
                start.emplace_back(h, w);
            }
        }
    }

    if (start.empty()) {
        for (int h = 0; h < H; ++h) {
            for (int w = 0; w < W; ++w) {
                cout << ans[h][w];
            }
            cout << '\n';
        }
        return 0;
    }

    queue<pair<pair<int, int>, pair<int, int>>> q;
    for (auto s : start) {
        q.push({s, s});
    }

    while (!q.empty()) {
        auto [pos, from] = q.front(); q.pop();
        for (auto nxt : get_next(pos, from)) {
            if (ans[nxt.first][nxt.second] == '.') {
                ans[nxt.first][nxt.second] = strByDir(nxt, pos);
                q.push({nxt, pos});
            }
        }
    }

    for (int h = 0; h < H; ++h) {
        for (int w = 0; w < W; ++w) {
            cout << ans[h][w];
        }
        cout << '\n';
    }

    return 0;
}

File: .__old/abc405/d.py
from collections import deque
H, W = map(int, input().split())
Grid = [list(input()) for _ in range(H)]

Directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]

def strByDir(a, b):
	d = (a[0] - b[0], a[1] - b[1])
	if d[0] == 1 and d[1] == 0:
		return '^'
	if d[0] == 0 and d[1] == 1:
		return '<'
	if d[0] == -1 and d[1] == 0:
		return 'v'
	if d[0] == 0 and d[1] == -1:
		return '>'

ans = Grid

start = []

for h in range(H):
	for w in range(W):
		if Grid[h][w] == 'E':
			start.append((h, w))

if not start:
	for h in range(H):
		print(''.join(ans[h]))
	exit(0)

def get_next(pos, _from):
	h, w = pos
	retval = []
	for dh, dw in Directions:
		if 0 <= h + dh < H and 0 <= w + dw < W:
			if h + dh == _from[0] and w + dw == _from[1]:
				continue
			if Grid[h + dh][w + dw] not in '#E':
				retval.append((h + dh, w + dw))
	return retval

queue = [(s, f) for s, f in zip(start, start)]
queue = deque(queue)

while queue:
	pos, _from = queue.popleft()
	for _next in get_next(pos, _from):
		if ans[_next[0]][_next[1]] == '.':
			ans[_next[0]][_next[1]] = strByDir(_next, pos)
			queue.append((_next, pos))

for h in range(H):
	print(''.join(ans[h]))
